### PR TITLE
Testing if Patch can take resourceVersion as a precondition

### PR DESCRIPTION
**Pull request by caesarxuchao**
Adding a test in test-cmd.sh to check if Patch can take resourceVersion as a precondition.

partially address #15416

@bgrant0607 @JanetKuo @lavalamp

### CODE REVIEW

**Comment by karlkfi** [comment]
make clean flake: #15424

File: hack/test-cmd.sh [hunk 1]
```diff
@@ -401,6 +401,21 @@ runTests() {
   # Post-condition: valid-pod POD has image kubernetes/pause
   kube::test::get_object_assert pods "{{range.items}}{{$image_field}}:{{end}}" 'kubernetes/pause:'
 
+  ## If resourceVersion is specified in the patch, it will be treated as a precondition, i.e., if the resourceVersion is different from that is stored in the server, the Patch should be rejected
+  ERROR_FILE="${KUBE_TEMP}/conflict-error"
+  resourceVersion=$(kubectl get "${kube_flags[@]}" pod valid-pod -o go-template='{{ .metadata.resourceVersion }}')
+  ((resourceVersion+=10))
+  # Command
+  kubectl patch "${kube_flags[@]}" pod valid-pod -p='{"spec":{"containers":[{"name": "kubernetes-serve-hostname", "image": "nginx"}]},"metadata":{"resourceVersion":"'$resourceVersion'"}}' 2> "${ERROR_FILE}" || true
```
Comment (JanetKuo) [inline hack/test-cmd.sh]: We can just put `! kubectl patch...` to imply that this command should fail (and then we don't need L410-417). 
Is there a test for the "resourceVersion is NOT different from the one stored in the server" case?

Comment (caesarxuchao) [inline hack/test-cmd.sh]: But I want to check the error message.

> Is there a test for the "resourceVersion is NOT different from the one stored in the server" case?
That's a good idea, I'll add one.

File: hack/test-cmd.sh [hunk 2]
```diff
@@ -428,6 +428,39 @@ runTests() {
   # Post-condition: valid-pod POD has image kubernetes/pause
   kube::test::get_object_assert pods "{{range.items}}{{$image_field}}:{{end}}" 'kubernetes/pause:'
 
+  ## If resourceVersion is specified in the patch, it will be treated as a precondition, i.e., if the resourceVersion is different from that is stored in the server, the Patch should be rejected
+  ERROR_FILE="${KUBE_TEMP}/conflict-error"
+  ## If the resourceVersion is the same as the one stored in the server, the patch will be applied.
+  # Command
+  # Needs to retry because other party may change the resource.
+  for count in $(seq 0 3); do
```
Comment (caesarxuchao) [inline hack/test-cmd.sh]: @JanetKuo, this tests patch would be accepted if the resourceVersion is correct. PTAL. Thank you.

File: hack/test-cmd.sh [hunk 3]
```diff
@@ -428,6 +428,39 @@ runTests() {
...
+    resourceVersion=$(kubectl get "${kube_flags[@]}" pod valid-pod -o go-template='{{ .metadata.resourceVersion }}')
+    kubectl patch "${kube_flags[@]}" pod valid-pod -p='{"spec":{"containers":[{"name": "kubernetes-serve-hostname", "image": "nginx"}]},"metadata":{"resourceVersion":"'$resourceVersion'"}}' 2> "${ERROR_FILE}" || true
+    if grep -q "the object has been modified" "${ERROR_FILE}"; then
+      kube::log::status "retry $1, error: $(cat ${ERROR_FILE})"
+      rm "${ERROR_FILE}"
```
Comment (JanetKuo) [inline hack/test-cmd.sh]: If you remove this file here you won't be able to access it in the next loop.

Comment (caesarxuchao) [inline hack/test-cmd.sh]: won't it be created at L438?

Comment (JanetKuo) [inline hack/test-cmd.sh]: Oh, yes. Never mind.

**Pull request merged**